eth-types/src/sign_types.rs

```diff
@@ -100,8 +106,8 @@ pub fn recover_pk(
     let pk = libsecp256k1::recover(&msg_hash, &signature, &recovery_id)?;
     let pk_be = pk.serialize();
     let pk_le = pk_bytes_swap_endianness(&pk_be[1..]);
-    let mut pk_bytes = secp256k1::Serialized::default();
-    pk_bytes.as_mut().copy_from_slice(&pk_le[..]);
+    let mut pk_bytes = secp256k1::Secp256k1Compressed::default();
+    pk_bytes.as_mut().copy_from_slice(&pk_le[..32]);
     let pk = Secp256k1Affine::from_bytes(&pk_bytes);
     ct_option_ok_or(pk, libsecp256k1::Error::InvalidPublicKey)
```

Comment:
Since `pk_le` is in uncompressed form `x || y`, `Secp256k1Affine::from_bytes(&pk_bytes)` always gives the point that `y` is even. To verify it's indeed even, we need check if `y` is the same value from `pk_le`, if not then we should negate the point.

```diff
-    ct_option_ok_or(pk, libsecp256k1::Error::InvalidPublicKey)
+    let pk = ct_option_ok_or(pk, libsecp256k1::Error::InvalidPublicKey)?;
+    debug_assert!(pk.y.to_bytes() == pk_le[32..] || (-pk).y.to_bytes() == pk_le[32..]);
+    if pk.y.to_bytes() == pk_le[32..] {
+        Ok(pk)
+    } else {
+        Ok(-pk)
+    }
```

Reply:
I think the second one is more readable and easy to follow although a bit more verbose. I'll go with it if it's ok for you.

Reply:
Makes sense, also second one should be more efficient (without calculating square root).